### DISCUSSION TITLE
How do I enable analysis in HistomicsUI?

### DISCUSSION THREAD

**Discussion by kwsp** [DigitalSlideArchive/digital_slide_archive#293] [Q&A] [answered]
![image](https://github.com/DigitalSlideArchive/digital_slide_archive/assets/30088608/d6f67250-8261-4f95-a0ce-808c5459916c)

I don't have the "analysis" button I seen in other screenshots, like this https://github.com/DigitalSlideArchive/digital_slide_archive/discussions/292

**Comment by manthey** [comment, marked as answer]
You need to add an analysis task to the system.  This can be done from the Girder interface (not the HistomicsUI interface).  Go to Collections -> Tasks -> Slicer CLI Web Tasks, you'll see an upload button that says "CLI" underneath it:
![image](https://github.com/DigitalSlideArchive/digital_slide_archive/assets/8781639/4aa539ce-f61f-4e30-a337-532fda84e449)
Click that and add the name of the docker image to add (e.g., "dsarchive/histomicstk:latest") and it will pull the docker image and add it to the available analyses.

**Reply by kwsp** [reply]
What do you mean by the "Girder interface"? Is that in the web UI or on the CLI?

**Reply by manthey** [reply]
The web UI at `<host>:<port>/` rather that `<host>:<port>/histomics`

**Reply by kwsp** [reply]
![image](https://github.com/DigitalSlideArchive/digital_slide_archive/assets/30088608/686bf00d-47bd-46e6-b217-316694632fad)

I think that's what I'm on. By the wait, I should mention I deployed HistomicsUI locally (without docker) following the [instructions here](https://github.com/DigitalSlideArchive/HistomicsUI#linux). The docker-compose config provided in the `digital_slide_archive` repository didn't work for me

**Reply by manthey** [reply]
The docker-compose would have set up everything; can you relate how it didn't work?

Otherwise, I don't know if you have the necessary parts for the analysis tasks -- it expects to be run as a user that is part of the docker group, as well as girder-worker to be started (which expects the rabbitmq service to be present).  The slicer-cli-web plugin settings can be used to specify a default folder for the analysis tasks, but that isn't actually required -- you can add analysis tasks through that plugin's settings, too.

**Reply by kwsp** [reply]
I see. For docker-compose, I followed the instructions here: https://github.com/DigitalSlideArchive/digital_slide_archive/tree/master/devops/dsa

The mongodb image keeps exiting with an exception and restarting. Here's the trace
```
pymongo.errors.ServerSelectionTimeoutError: mongodb:27017: [Errno -2] Name or service not known, Timeout: 20.0s, Topology Description: <TopologyDescription id: 64ff705a94872311118a7a67, topology_type: Single, servers: [<ServerDescription ('mongodb', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('mongodb:27017: [Errno -2] Name or service not known')>]>
```

This is the full log of the restart loop:

```
girder_1     | The user `ubuntu' is already a member of `ubuntu'.
girder_1     | The user `ubuntu' is already a member of `root'.
girder_1     | net.ipv4.conf.eth0.route_localnet = 1
girder_1     | ==== Pre-Provisioning ===
girder_1     | Provision file date: Fri Sep  8 19:29:29 2023; size: 26781
girder_1     | ==== Provisioning ===
girder_1     | Provision file date: Fri Sep  8 19:29:29 2023; size: 26781
girder_1     | Running in mode: development
girder_1     | Connecting to MongoDB: mongodb://mongodb:27017/girder?socketTimeoutMS=3600000
girder_1     | Traceback (most recent call last):
girder_1     |   File "/opt/digital_slide_archive/devops/dsa/provision.py", line 625, in <module>
girder_1     |     configureServer()
girder_1     |   File "/opt/girder/girder/utility/server.py", line 80, in configureServer
girder_1     |     api_main.addApiToNode(root)
girder_1     |   File "/opt/girder/girder/api/api_main.py", line 20, in addApiToNode
girder_1     |     _addV1ToNode(node.api)
girder_1     |   File "/opt/girder/girder/api/api_main.py", line 30, in _addV1ToNode
girder_1     |     node.v1.assetstore = assetstore.Assetstore()
girder_1     |   File "/opt/girder/girder/api/v1/assetstore.py", line 23, in __init__
girder_1     |     self._model = AssetstoreModel()
girder_1     |   File "/opt/girder/girder/models/model_base.py", line 76, in __call__
girder_1     |     cls._instance = super().__call__(*args, **kwargs)
girder_1     |   File "/opt/girder/girder/models/model_base.py", line 109, in __init__
girder_1     |     self.reconnect()
girder_1     |   File "/opt/girder/girder/models/model_base.py", line 116, in reconnect
girder_1     |     db_connection = getDbConnection()
girder_1     |   File "/opt/girder/girder/models/__init__.py", line 97, in getDbConnection
girder_1     |     client.server_info()
girder_1     |   File "/opt/venv/lib/python3.9/site-packages/pymongo/mongo_client.py", line 2086, in server_info
girder_1     |     return self.admin.command(
girder_1     |   File "/opt/venv/lib/python3.9/site-packages/pymongo/database.py", line 827, in command
girder_1     |     with self.__client._socket_for_reads(read_preference, session) as (sock_info, secondary_ok):
girder_1     |   File "/.pyenv/versions/3.9.17/lib/python3.9/contextlib.py", line 119, in __enter__
girder_1     |     return next(self.gen)
girder_1     |   File "/opt/venv/lib/python3.9/site-packages/pymongo/mongo_client.py", line 1478, in _socket_for_reads
girder_1     |     server = self._select_server(read_preference, session)
girder_1     |   File "/opt/venv/lib/python3.9/site-packages/pymongo/mongo_client.py", line 1436, in _select_server
girder_1     |     server = topology.select_server(server_selector)
girder_1     |   File "/opt/venv/lib/python3.9/site-packages/pymongo/topology.py", line 250, in select_server
girder_1     |     return random.choice(self.select_servers(selector, server_selection_timeout, address))
girder_1     |   File "/opt/venv/lib/python3.9/site-packages/pymongo/topology.py", line 211, in select_servers
girder_1     |     server_descriptions = self._select_servers_loop(selector, server_timeout, address)
girder_1     |   File "/opt/venv/lib/python3.9/site-packages/pymongo/topology.py", line 226, in _select_servers_loop
girder_1     |     raise ServerSelectionTimeoutError(
girder_1     | pymongo.errors.ServerSelectionTimeoutError: mongodb:27017: [Errno -2] Name or service not known, Timeout: 20.0s, Topology Description: <TopologyDescription id: 64ff70c707ab27c7f607071c, topology_type: Single, servers: [<ServerDescription ('mongodb', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('mongodb:27017: [Errno -2] Name or service not known')>]>
dsa_girder_1 exited with code 1
```

**Reply by manthey** [reply]
It sounds like the mongodb container failed to start.  If you do `docker logs dsa-mongodb-1` (that last name might be slightly different depending on your docker version), you'll see whatever went wrong with mongo.  If you are running in a virtualized linux, you might not have avx instructions enabled, in which case Mongo >= 5 won't run and you have to either enable avx instructions or specify the mongo:4 image for the container.

**Reply by kwsp** [reply]
`docker logs` showed this error: `{"t":{"$date":"2023-09-08T22:36:21.218+00:00"},"s":"E",  "c":"STORAGE",  "id":22312,   "ctx":"initandlisten","msg":"Error creating journal directory","attr":{"directory":"/data/db/journal","error":"boost::filesystem::create_directory: Permission denied [system:13]: \"/data/db/journal\""}}`

Any ideas why this is happening? I didn't modify the docker-compose config file or the mongo image, so it should just work. Ubuntu is running on bare metal.

**Reply by kwsp** [reply]
The mongo container restarts about every 7 seconds, so I can't really exec into the container to check the permission issue.

**Reply by manthey** [reply]
Hmm...  The default deployment mounts the `devops/dsa/db` directory as the internal container `/data/db` directory, so you should be able to do `ls -l devops/dsa/db` and see if any mongo created files are owned by your current user.  If you tried running docker compose as a different user, this could cause such a problem, in which case chown-ing the files in the db directory to your current user might fix it.

**Reply by kwsp** [reply]
Looks like the mongo container doesnt have the permission to do anything in the `/data/db` directory. I see its the `devlops/dsa/db` directory, and it is owned by the current user. I wonder if `DSA_USER=$(id -u):$(id -g)` is doing something funky.

**Reply by kwsp** [reply]
So the `devops/dsa/db` directory is owned by me, and I'm running docker-compose as my user, and I provided `DSA_USER=$(id -u):$(id -g)`, yet mongo doesn't have permission to write to `/data/db`. Do you have any other ideas on what might've gone wrong?

**Reply by kwsp** [reply]
@manthey I tried docker-compose again in a clean Ubuntu install and it worked. Still not sure how to fix the problem I saw before.
